src/futures/spi.rs

```diff
@@ -0,0 +1,295 @@
+///
+/// - Due to how full duplex SPI works each `read` call must be preceded by a `write` call.
+///
+/// - `read` calls only return the data received with the last `write` call.
+/// Previously received data is discarded
+///
+/// - Data is only guaranteed to be clocked out when the `read` call succeeds.
+/// The slave select line shouldn't be released before that.
+///
+/// - Some SPIs can work with 8-bit *and* 16-bit words. You can overload this trait with different
+/// `Word` types to allow operation in both modes.
+pub trait FullDuplex<Word> {
+    /// An enumeration of SPI errors
+    type Error;
+
+    /// The future associated with the `read` method.
+    type ReadFuture<'a>: Future<Output=Result<Word, Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// The future associated with the `write` method.
+    type WriteFuture<'a>: Future<Output=Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// Reads the word stored in the shift register
+    ///
+    /// **NOTE** A word must be sent to the slave before attempting to call this
+    /// method.
+    fn read<'a>(&'a mut self) -> Self::ReadFuture<'a>;
+
+    /// Writes a word to the slave
+    fn write<'a>(&'a mut self, word: Word) -> Self::WriteFuture<'a>;
+}
+
+/// Clock polarity
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum Polarity {
+    /// Clock signal low when idle
+    IdleLow,
+    /// Clock signal high when idle
+    IdleHigh,
+}
+
+/// Clock phase
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum Phase {
+    /// Data in "captured" on the first clock transition
+    CaptureOnFirstTransition,
+    /// Data in "captured" on the second clock transition
+    CaptureOnSecondTransition,
+}
+
+/// SPI mode
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct Mode {
+    /// Clock polarity
+    pub polarity: Polarity,
+    /// Clock phase
+    pub phase: Phase,
+}
+
+/// Helper for CPOL = 0, CPHA = 0
+pub const MODE_0: Mode = Mode {
+    polarity: Polarity::IdleLow,
+    phase: Phase::CaptureOnFirstTransition,
+};
+
+/// Helper for CPOL = 0, CPHA = 1
+pub const MODE_1: Mode = Mode {
+    polarity: Polarity::IdleLow,
+    phase: Phase::CaptureOnSecondTransition,
+};
+
+/// Helper for CPOL = 1, CPHA = 0
+pub const MODE_2: Mode = Mode {
+    polarity: Polarity::IdleHigh,
+    phase: Phase::CaptureOnFirstTransition,
+};
+
+/// Helper for CPOL = 1, CPHA = 1
+pub const MODE_3: Mode = Mode {
+    polarity: Polarity::IdleHigh,
+    phase: Phase::CaptureOnSecondTransition,
+};
+
+/// Async transfer
+pub trait Transfer<Word: 'static> {
+    /// Error type
+    type Error;
+
+    /// Associated future for the `transfer` method.
+    type TransferFuture<'a>: Future<Output = Result<&'a [Word], Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// Writes `words` to the slave. Returns the `words` received from the slave
+    fn transfer<'a>(&'a mut self, words: &'a mut [Word]) -> Self::TransferFuture<'a>
+    where
```

Comment:
Why `Self::Error: 'a`?

Reply:
Oops, that was left over from some stuff before.